packages/compass-aggregations/src/modules/pipeline-builder/pipeline-builder.ts

```diff
@@ -241,4 +226,57 @@ export class PipelineBuilder {
       force
     );
   }
-}
\ No newline at end of file
+
+  /**
+   * Returns true if previous preview request was done for the same pipeline for
+   * a specific stage
+   */
+  isLastStagePreviewEqual(
+    idx: number,
+    pipeline: Document[] = this.getPipelineFromStages(
+      this.stages.slice(0, idx + 1)
+    )
+  ) {
+    return this.previewManager.isLastPipelineEqual(idx, pipeline);
+  }
+
+  /**
+   * Request preview for current pipeline source
+   */
+  getPreviewForPipeline(
+    namespace: string,
+    options: PreviewOptions,
+    filterOutputStage = false
+  ): Promise<Document[]> {
+    // For preview we ignore $out/$merge stage.
+    const pipeline = [...this.getPipelineFromSource()];
+    if (filterOutputStage && isLastStageOutputStage(pipeline)) {
+      pipeline.pop();
+    }
+    return this.previewManager.getPreviewForStage(
+      FULL_PIPELINE_PREVIEW_ID,
+      namespace,
+      pipeline,
+      options
+    );
+  }
+
+  /**
+   * Returns true if previous preview request was done for the same pipeline for
+   * the whole pipeline
+   */
+  isLastPipelinePreviewEqual(
+    pipeline: Document[] = this.getPipelineFromSource(),
+    filterOutputStage = false
+  ) {
+    pipeline = [...pipeline];
+    // For preview we ignore $out/$merge stage.
+    if (filterOutputStage && isLastStageOutputStage(pipeline)) {
+      pipeline.pop();
+    }
+    return this.previewManager.isLastPipelineEqual(
+      FULL_PIPELINE_PREVIEW_ID,
+      pipeline
+    );
+  }
```

Comment:
@mabaasit we discussed this already I think and were on the same page, but wanted to highlight that based on these changes to the design we are making last minute the initial idea of only exposing preview manager through the pipeline builder doesn't make much sense anymore, we probably want to separate them again and directly use in the actions instead of through the pipelineBuilder as more and more we are using the preview methods with last successful parsing of the pipeline and not the actual current state of the pipeline.

I am keeping it as-is right now because this would be a much bigger change and requires some additional thinking through and this seems to be good enough for now